File: backend/app/schemas/feed.py
from pydantic import BaseModel, Field
from typing import Optional
from uuid import UUID
from datetime import datetime


class PollOptionCreate(BaseModel):
    text: str = Field(..., max_length=500)


class PollCreate(BaseModel):
    question: str = Field(..., max_length=500)
    options: list[PollOptionCreate] = Field(..., min_length=2, max_length=10)
    ends_at: Optional[datetime] = None
    is_anonymous: bool = False


class PollOptionResponse(BaseModel):
    id: UUID
    text: str
    votes_count: int
    percentage: float = 0.0
    has_voted: bool = False

    model_config = {"from_attributes": True}


class PollResponse(BaseModel):
    id: UUID
    question: str
    ends_at: Optional[datetime]
    is_anonymous: bool
    total_votes: int
    options: list[PollOptionResponse]
    has_voted: bool = False
    is_expired: bool = False

    model_config = {"from_attributes": True}


class PostCreate(BaseModel):
    content: Optional[str] = None
    image_urls: Optional[list[str]] = None
    video_url: Optional[str] = None
    audio_url: Optional[str] = None
    gif_url: Optional[str] = None
    document_url: Optional[str] = None
    document_name: Optional[str] = None
    location_name: Optional[str] = None
    location_lat: Optional[float] = None
    location_lng: Optional[float] = None
    feeling_type: Optional[str] = None
    feeling_text: Optional[str] = None
    background_style: Optional[str] = None
    background_image_url: Optional[str] = None
    aspect_ratio: Optional[str] = None
    post_type: str = Field(default="text", pattern=r"^(text|image|video|audio|gif|poll|document|shared|quote)$")
    privacy: Optional[str] = Field(default=None, pattern=r"^(everyone|friends|only_me)$")
    is_draft: bool = False
    is_scheduled: bool = False
    scheduled_at: Optional[datetime] = None
    shared_post_id: Optional[str] = None
    quote_text: Optional[str] = None
    cross_posted_from: Optional[str] = None
    poll: Optional[PollCreate] = None
    hashtags: Optional[list[str]] = None


class PostUpdate(BaseModel):
    content: Optional[str] = None
    image_urls: Optional[list[str]] = None
    video_url: Optional[str] = None
    audio_url: Optional[str] = None
    gif_url: Optional[str] = None
    document_url: Optional[str] = None
    document_name: Optional[str] = None
    location_name: Optional[str] = None
    location_lat: Optional[float] = None
    location_lng: Optional[float] = None
    feeling_type: Optional[str] = None
    feeling_text: Optional[str] = None
    background_style: Optional[str] = None
    background_image_url: Optional[str] = None
    aspect_ratio: Optional[str] = None
    privacy: Optional[str] = Field(None, pattern=r"^(everyone|friends|only_me)$")
    is_archived: Optional[bool] = None
    is_draft: Optional[bool] = None
    scheduled_at: Optional[datetime] = None
    quote_text: Optional[str] = None


class PostAuthor(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    is_verified: bool

    model_config = {"from_attributes": True}


class PostResponse(BaseModel):
    id: UUID
    user_id: UUID
    content: Optional[str]
    image_urls: Optional[list[str]] = None
    video_url: Optional[str]
    audio_url: Optional[str]
    gif_url: Optional[str]
    document_url: Optional[str]
    document_name: Optional[str]
    location_name: Optional[str]
    location_lat: Optional[float]
    location_lng: Optional[float]
    feeling_type: Optional[str]
    feeling_text: Optional[str]
    background_style: Optional[str] = None
    background_image_url: Optional[str] = None
    aspect_ratio: Optional[str] = None
    post_type: str
    privacy: str
    is_pinned: bool
    is_hidden: bool
    is_archived: bool
    is_draft: bool
    is_scheduled: bool
    scheduled_at: Optional[datetime]
    shared_post_id: Optional[UUID]
    quote_text: Optional[str]
    cross_posted_from: Optional[str]
    repost_count: int
    likes_count: int
    comments_count: int
    shares_count: int
    trending_score: float
    author: Optional[PostAuthor] = None
    shared_post: Optional["PostResponse"] = None
    poll: Optional[PollResponse] = None
    is_liked: bool = False
    is_saved: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class FeedResponse(BaseModel):
    posts: list[PostResponse]
    next_cursor: Optional[str] = None
    has_more: bool


class FeedPositionResponse(BaseModel):
    user_id: UUID
    feed_type: str
    last_post_id: Optional[UUID]
    scroll_position: int
    updated_at: datetime

    model_config = {"from_attributes": True}


class FeedPositionUpdate(BaseModel):
    feed_type: str = "home"
    last_post_id: Optional[str] = None
    scroll_position: int = 0


class FeedFilterOptions(BaseModel):
    feed_type: str = "home"
    privacy_filter: Optional[str] = None
    sort_by: str = "latest"


class RepostCreate(BaseModel):
    content: Optional[str] = None


class QuoteCreate(BaseModel):
    quote_text: str = Field(..., max_length=500)
    content: Optional[str] = None

File: frontend/src/types/feed.ts
export interface PostAuthor {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  is_verified: boolean;
}

export interface PollOption {
  id: string;
  text: string;
  votes_count: number;
  percentage: number;
  has_voted: boolean;
}

export interface Poll {
  id: string;
  question: string;
  ends_at: string | null;
  is_anonymous: boolean;
  total_votes: number;
  options: PollOption[];
  has_voted: boolean;
  is_expired: boolean;
}

export interface PollCreate {
  question: string;
  options: { text: string }[];
  ends_at?: string;
  is_anonymous?: boolean;
}

export interface Post {
  id: string;
  user_id: string;
  content: string | null;
  image_urls: string[] | null;
  video_url: string | null;
  audio_url: string | null;
  gif_url: string | null;
  document_url: string | null;
  document_name: string | null;
  location_name: string | null;
  location_lat: number | null;
  location_lng: number | null;
  feeling_type: string | null;
  feeling_text: string | null;
  background_style: string | null;
  background_image_url: string | null;
  aspect_ratio: string | null;
  post_type: PostType;
  privacy: "everyone" | "friends" | "close_friends" | "followers" | "friends_followers" | "only_me";
  is_pinned: boolean;
  is_hidden: boolean;
  is_archived: boolean;
  is_draft: boolean;
  is_scheduled: boolean;
  scheduled_at: string | null;
  shared_post_id: string | null;
  quote_text: string | null;
  cross_posted_from: string | null;
  repost_count: number;
  likes_count: number;
  comments_count: number;
  shares_count: number;
  trending_score: number;
  author: PostAuthor | null;
  shared_post: Post | null;
  poll: Poll | null;
  is_liked: boolean;
  is_saved: boolean;
  created_at: string;
  updated_at: string;
}

export type PostType = "text" | "image" | "video" | "audio" | "gif" | "poll" | "document" | "shared" | "quote";

export interface PostCreate {
  content?: string;
  image_urls?: string[];
  video_url?: string;
  audio_url?: string;
  gif_url?: string;
  document_url?: string;
  document_name?: string;
  location_name?: string;
  location_lat?: number;
  location_lng?: number;
  feeling_type?: string;
  feeling_text?: string;
  background_style?: string;
  background_image_url?: string;
  aspect_ratio?: string;
  post_type?: PostType;
  privacy?: "everyone" | "friends" | "close_friends" | "followers" | "friends_followers" | "only_me";
  is_draft?: boolean;
  is_scheduled?: boolean;
  scheduled_at?: string;
  shared_post_id?: string;
  quote_text?: string;
  cross_posted_from?: string;
  poll?: PollCreate;
  hashtags?: string[];
}

export interface PostUpdate {
  content?: string;
  image_urls?: string[];
  video_url?: string;
  audio_url?: string;
  gif_url?: string;
  document_url?: string;
  document_name?: string;
  location_name?: string;
  location_lat?: number;
  location_lng?: number;
  feeling_type?: string;
  feeling_text?: string;
  background_style?: string;
  background_image_url?: string;
  aspect_ratio?: string;
  privacy?: "everyone" | "friends" | "close_friends" | "followers" | "friends_followers" | "only_me";
  is_archived?: boolean;
  is_draft?: boolean;
  scheduled_at?: string;
  quote_text?: string;
}

export interface FeedResponse {
  posts: Post[];
  next_cursor: string | null;
  has_more: boolean;
}

export interface FeedPosition {
  user_id: string;
  feed_type: string;
  last_post_id: string | null;
  scroll_position: number;
  updated_at: string;
}

export interface FeedPositionUpdate {
  feed_type: string;
  last_post_id?: string;
  scroll_position?: number;
}

export type FeedType = "home" | "following" | "friends" | "trending" | "suggested";

export type FeedSortBy = "latest" | "trending" | "popular";

export interface FeedFilters {
  feed_type: FeedType;
  sort_by: FeedSortBy;
  privacy_filter?: "everyone" | "friends" | "close_friends" | "followers" | "friends_followers" | "only_me";
}

export const POST_TYPES = [
  { value: "text", label: "Text", icon: "Type" },
  { value: "image", label: "Photo", icon: "Image" },
  { value: "video", label: "Video", icon: "Video" },
  { value: "poll", label: "Poll", icon: "BarChart3" },
] as const;

export const FEELING_TYPES = [
  { value: "happy", label: "Happy", emoji: "😊" },
  { value: "sad", label: "Sad", emoji: "😢" },
  { value: "excited", label: "Excited", emoji: "🎉" },
  { value: "angry", label: "Angry", emoji: "😠" },
  { value: "love", label: "In Love", emoji: "❤️" },
  { value: "grateful", label: "Grateful", emoji: "🙏" },
  { value: "tired", label: "Tired", emoji: "😴" },
  { value: "blessed", label: "Blessed", emoji: "✨" },
  { value: "thinking", label: "Thinking", emoji: "🤔" },
  { value: "celebrating", label: "Celebrating", emoji: "🥳" },
] as const;

export const FEED_TABS = [
  { key: "home", label: "Home" },
  { key: "following", label: "Following" },
  { key: "friends", label: "Friends" },
  { key: "trending", label: "Trending" },
  { key: "suggested", label: "For You" },
] as const;

export const FEED_SORT_OPTIONS = [
  { value: "latest", label: "Latest" },
  { value: "trending", label: "Trending" },
  { value: "popular", label: "Popular" },
] as const;

export const POST_VISIBILITY_OPTIONS = [
  { value: "everyone", label: "Everyone" },
  { value: "friends", label: "Friends" },
  { value: "close_friends", label: "Close Friends" },
  { value: "followers", label: "Followers" },
  { value: "friends_followers", label: "Friends & Followers" },
  { value: "only_me", label: "Only Me" },
] as const;
